File: apps/api/app/schemas/content.py
"""教案 / 学案结构化内容模型。

替代旧的 8-block 体系，用明确的教案/学案结构化模型表达教学内容。
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# 公共类型
# ---------------------------------------------------------------------------

SectionStatus = Literal["confirmed", "pending"]

Scene = Literal["public_school", "tutor", "institution"]

LessonType = Literal["lesson_plan", "study_guide", "both"]

LessonCategory = Literal["new", "review", "exercise", "comprehensive"]


class TeachingObjective(BaseModel):
    """教学目标（三维目标或核心素养）。"""

    dimension: Literal["knowledge", "ability", "emotion"] = "knowledge"
    content: str = ""


class KeyPoints(BaseModel):
    """教学重难点。"""

    key_points: list[str] = Field(default_factory=list)
    difficulties: list[str] = Field(default_factory=list)


class TeachingProcessStep(BaseModel):
    """教学过程的一个环节。"""

    phase: str = ""
    duration: int = Field(default=5, description="分钟")
    teacher_activity: str = ""
    student_activity: str = ""
    design_intent: str = ""
    status: SectionStatus = "pending"


class AssessmentItem(BaseModel):
    """学案中的测评/练习题。"""

    level: Literal["A", "B", "C", "D"] = "A"
    item_type: Literal["choice", "fill_blank", "short_answer"] = Field(
        default="short_answer", alias="itemType"
    )
    prompt: str = ""
    options: list[str] = Field(default_factory=list)
    answer: str = ""
    analysis: str = ""

    model_config = {"populate_by_name": True}


# ---------------------------------------------------------------------------
# 教案内容模型
# ---------------------------------------------------------------------------


class LessonPlanHeader(BaseModel):
    """教案基本信息。"""

    title: str = ""
    subject: str = ""
    grade: str = ""
    class_hour: int = Field(default=1, alias="classHour")
    lesson_category: LessonCategory = Field(default="new", alias="lessonCategory")
    teacher: str = ""

    model_config = {"populate_by_name": True}


class LessonPlanContent(BaseModel):
    """教案完整内容。"""

    doc_type: Literal["lesson_plan"] = "lesson_plan"
    header: LessonPlanHeader = Field(default_factory=LessonPlanHeader)
    objectives: list[TeachingObjective] = Field(default_factory=list)
    objectives_status: SectionStatus = "pending"
    key_points: KeyPoints = Field(default_factory=KeyPoints)
    key_points_status: SectionStatus = "pending"
    preparation: list[str] = Field(default_factory=list)
    preparation_status: SectionStatus = "pending"
    teaching_process: list[TeachingProcessStep] = Field(default_factory=list)
    teaching_process_status: SectionStatus = "pending"
    board_design: str = ""
    board_design_status: SectionStatus = "pending"
    reflection: str = ""
    reflection_status: SectionStatus = "pending"


# ---------------------------------------------------------------------------
# 学案内容模型
# ---------------------------------------------------------------------------


class StudyGuideHeader(BaseModel):
    """学案基本信息。"""

    title: str = ""
    subject: str = ""
    grade: str = ""
    class_name: str = Field(default="", alias="className")
    student_name: str = Field(default="", alias="studentName")
    date: str = ""

    model_config = {"populate_by_name": True}


class LearningProcess(BaseModel):
    """学案学习流程（三段式）。"""

    self_study: list[AssessmentItem] = Field(default_factory=list, alias="selfStudy")
    collaboration: list[AssessmentItem] = Field(default_factory=list)
    presentation: list[AssessmentItem] = Field(default_factory=list)

    model_config = {"populate_by_name": True}


class StudyGuideContent(BaseModel):
    """学案完整内容。"""

    doc_type: Literal["study_guide"] = "study_guide"
    header: StudyGuideHeader = Field(default_factory=StudyGuideHeader)
    learning_objectives: list[str] = Field(default_factory=list)
    learning_objectives_status: SectionStatus = "pending"
    key_difficulties: list[str] = Field(default_factory=list)
    key_difficulties_status: SectionStatus = "pending"
    prior_knowledge: list[str] = Field(default_factory=list)
    prior_knowledge_status: SectionStatus = "pending"
    learning_process: LearningProcess = Field(default_factory=LearningProcess)
    self_study_status: SectionStatus = "pending"
    collaboration_status: SectionStatus = "pending"
    presentation_status: SectionStatus = "pending"
    assessment: list[AssessmentItem] = Field(default_factory=list)
    assessment_status: SectionStatus = "pending"
    extension: list[AssessmentItem] = Field(default_factory=list)
    extension_status: SectionStatus = "pending"
    self_reflection: str = ""
    self_reflection_status: SectionStatus = "pending"


# ---------------------------------------------------------------------------
# 联合类型
# ---------------------------------------------------------------------------

DocumentContent = LessonPlanContent | StudyGuideContent


def is_lesson_plan(doc: DocumentContent) -> bool:
    return getattr(doc, "doc_type", None) == "lesson_plan"


def is_study_guide(doc: DocumentContent) -> bool:
    return getattr(doc, "doc_type", None) == "study_guide"


def create_empty_lesson_plan(
    *,
    subject: str = "",
    grade: str = "",
    topic: str = "",
    class_hour: int = 1,
    lesson_category: LessonCategory = "new",
) -> LessonPlanContent:
    """创建空的教案模板。"""
    return LessonPlanContent(
        header=LessonPlanHeader(
            title=topic,
            subject=subject,
            grade=grade,
            class_hour=class_hour,
            lesson_category=lesson_category,
        ),
        objectives=[],
        objectives_status="pending",
        key_points=KeyPoints(),
        key_points_status="pending",
        preparation=[],
        preparation_status="pending",
        teaching_process=[],
        teaching_process_status="pending",
        board_design="",
        board_design_status="pending",
        reflection="",
        reflection_status="pending",
    )


def create_empty_study_guide(
    *,
    subject: str = "",
    grade: str = "",
    topic: str = "",
) -> StudyGuideContent:
    """创建空的学案模板。"""
    return StudyGuideContent(
        header=StudyGuideHeader(
            title=topic,
            subject=subject,
            grade=grade,
        ),
        learning_objectives=[],
        learning_objectives_status="pending",
        key_difficulties=[],
        key_difficulties_status="pending",
        prior_knowledge=[],
        prior_knowledge_status="pending",
        learning_process=LearningProcess(),
        self_study_status="pending",
        collaboration_status="pending",
        presentation_status="pending",
        assessment=[],
        assessment_status="pending",
        extension=[],
        extension_status="pending",
        self_reflection="",
        self_reflection_status="pending",
    )

File: packages/shared-types/src/content.ts
/**
 * 教案 / 学案结构化内容类型定义。
 *
 * 与后端 schemas/content.py 对齐。
 */

// ---------------------------------------------------------------------------
// 公共枚举
// ---------------------------------------------------------------------------

export type SectionStatus = 'confirmed' | 'pending';

export type Scene = 'public_school' | 'tutor' | 'institution';

export type LessonType = 'lesson_plan' | 'study_guide' | 'both';

export type LessonCategory = 'new' | 'review' | 'exercise' | 'comprehensive';

// ---------------------------------------------------------------------------
// 公共类型
// ---------------------------------------------------------------------------

export interface TeachingObjective {
  dimension: 'knowledge' | 'ability' | 'emotion';
  content: string;
}

export interface KeyPoints {
  keyPoints: string[];
  difficulties: string[];
}

export interface TeachingProcessStep {
  phase: string;
  duration: number;
  teacher_activity: string;
  student_activity: string;
  design_intent: string;
  status?: SectionStatus;
}

export interface AssessmentItem {
  level: 'A' | 'B' | 'C' | 'D';
  itemType: 'choice' | 'fill_blank' | 'short_answer';
  prompt: string;
  options: string[];
  answer: string;
  analysis: string;
}

// ---------------------------------------------------------------------------
// 教案
// ---------------------------------------------------------------------------

export interface LessonPlanHeader {
  title: string;
  subject: string;
  grade: string;
  classHour: number;
  lessonCategory: LessonCategory;
  teacher: string;
}

export interface LessonPlanContent {
  doc_type: 'lesson_plan';
  header: LessonPlanHeader;
  objectives: TeachingObjective[];
  objectives_status: SectionStatus;
  key_points: KeyPoints;
  key_points_status: SectionStatus;
  preparation: string[];
  preparation_status: SectionStatus;
  teaching_process: TeachingProcessStep[];
  teaching_process_status: SectionStatus;
  board_design: string;
  board_design_status: SectionStatus;
  reflection: string;
  reflection_status: SectionStatus;
}

// ---------------------------------------------------------------------------
// 学案
// ---------------------------------------------------------------------------

export interface StudyGuideHeader {
  title: string;
  subject: string;
  grade: string;
  className: string;
  studentName: string;
  date: string;
}

export interface LearningProcess {
  selfStudy: AssessmentItem[];
  collaboration: AssessmentItem[];
  presentation: AssessmentItem[];
}

export interface StudyGuideContent {
  doc_type: 'study_guide';
  header: StudyGuideHeader;
  learning_objectives: string[];
  learning_objectives_status: SectionStatus;
  key_difficulties: string[];
  key_difficulties_status: SectionStatus;
  prior_knowledge: string[];
  prior_knowledge_status: SectionStatus;
  learning_process: LearningProcess;
  self_study_status: SectionStatus;
  collaboration_status: SectionStatus;
  presentation_status: SectionStatus;
  assessment: AssessmentItem[];
  assessment_status: SectionStatus;
  extension: AssessmentItem[];
  extension_status: SectionStatus;
  self_reflection: string;
  self_reflection_status: SectionStatus;
}

// ---------------------------------------------------------------------------
// 联合类型 & 类型守卫
// ---------------------------------------------------------------------------

export type DocumentContent = LessonPlanContent | StudyGuideContent;

export function isLessonPlan(content: DocumentContent): content is LessonPlanContent {
  return content.doc_type === 'lesson_plan';
}

export function isStudyGuide(content: DocumentContent): content is StudyGuideContent {
  return content.doc_type === 'study_guide';
}

// ---------------------------------------------------------------------------
// Section 元信息
// ---------------------------------------------------------------------------

export interface SectionInfo {
  name: string;
  title: string;
  status: SectionStatus;
}
